File: src/ecommerce-frontend/src/app/core/api/api-response.ts
import { HttpErrorResponse } from '@angular/common/http';

export interface ApiError {
  code: string;
  message: string;
  details: unknown;
  traceId: string | null;
}

export interface ApiResponse<T> {
  success: boolean;
  data: T | null;
  error: ApiError | null;
}

export class ApiFailure extends Error {
  constructor(
    readonly code: string,
    message: string,
    readonly traceId: string | null = null,
    readonly details: unknown = null,
  ) {
    super(message);
    this.name = 'ApiFailure';
  }
}

export function unwrap<T>(response: ApiResponse<T>): T {
  if (!response.success) {
    throw fromApiError(response.error);
  }
  return response.data as T;
}

export function toApiFailure(error: unknown): ApiFailure {
  if (error instanceof ApiFailure) {
    return error;
  }
  if (error instanceof HttpErrorResponse) {
    const body = error.error as Partial<ApiResponse<unknown>> | null;
    if (body && typeof body === 'object' && body.error) {
      return fromApiError(body.error);
    }
    if (error.status === 0) {
      return new ApiFailure('client.network_error', 'The server could not be reached');
    }
    return new ApiFailure(`common.http_${error.status}`, error.statusText || 'Request failed');
  }
  return new ApiFailure('client.unknown_error', 'Something went wrong');
}

function fromApiError(error: ApiError | null): ApiFailure {
  return new ApiFailure(
    error?.code ?? 'client.unknown_error',
    error?.message ?? 'Request failed',
    error?.traceId ?? null,
    error?.details ?? null,
  );
}

File: src/Ecommerce.ApiDefaults/ApiResponse.cs
using System.Diagnostics;

namespace Ecommerce.ApiDefaults;

public sealed record ApiError(string Code, string Message, object? Details, string? TraceId);

public sealed record ApiResponse<T>(bool Success, T? Data, ApiError? Error);

public sealed record ApiResponse(bool Success, object? Data, ApiError? Error)
{
    public static ApiResponse<T> Ok<T>(T data) => new(true, data, null);

    public static ApiResponse Empty() => new(true, null, null);

    public static ApiResponse Fail(string code, string message, object? details = null) =>
        new(false, null, new ApiError(code, message, details, Activity.Current?.Id));
}
